### DISCUSSION TITLE
Java Issues when trying to run Freeplane Portable 1.10.2

### DISCUSSION THREAD

**Discussion by sifran-github** [freeplane/freeplane#462] [Help] [answered]
I cant seem to run Freeplane Portable (installed from: FreeplanePortable-1.10.2.paf.exe).
I think the latest version 10 created some issues with the java compatability.

I started by having the latest java on my computer, 18 (1.8.0_333).

I ran FreeplanePortable-1.10.2.paf.exe and extracted the files, and then when I tried to run FreeplanePortable.exe, I got the following error:
Java Virtual Machine Launcher
Error: Could not create the Java Virtual Machine.
Error: A fatal exception has occurred. Program will exit.

I tried everything from here:
https://www.happycoders.eu/java/how-to-switch-multiple-java-versions-windows/
https://windowsreport.com/java-virtual-machine-fatal-error/
https://github.com/freeplane/freeplane/discussions/275

I tried to set JAVA_HOME, FREEPLANE_JAVA_HOME env variables.
Then I tried to install java 17 (why i cant find JRE 17?! i can only find the full jdk...) and it still doesn't work. same error.

Finally, i tried to download the regular installation without the java (Freeplane-Setup-1.10.2.exe).
the result is:
When setting JAVA_HOME to C:\Program Files (x86)\Java\jre1.8.0_333: Both version do NOT work.
when setting JAVA_HOME to C:\Program Files\Java\jdk-17.0.3.1: The regular freeplane works, but the portable version is still throwing the same error "Could not create the Java Virtual Machine.".

What is the proper way to run Freeplane 1.10.2 with java 18?
Maybe there is an issue with the Portable version?

**Comment by sifran-github** [comment]
I think that maybe the portable version ignores my evn variable settings for some reason, and keep finding my latest java version (18) which cannot run freeplane for some reason?

even if i remove java completely from the path, and try to run both versions:

The regular version throws:
"This application requires a Java Runtime Environment 11 - 17"

And the portable version throws the same error as usual:
"Could not create the Java Virtual Machine."

**Comment by dpolivaev** [comment]
1. 1.8.0_333 is Java 8 update 333
2. Freeplane 1.10 runs neither with Java 8 nor with Java 18.
3. To run the portable version you need to install https://sourceforge.net/projects/portableapps/files/OpenJDK%20JRE%20Portable/OpenJDKJRE64_15.0.2_Build_7.paf.exe/download as your portable app

**Comment by dpolivaev** [comment]
By the way, searching https://github.com/freeplane/freeplane/search?q=portable&type=discussions finds https://github.com/freeplane/freeplane/discussions/365 where this issue has already been discussed

**Comment by sifran-github** [comment, marked as answer]
Sorry. i searched google and freeplane forum for the exact error ("Could not create the Java Virtual Machine") and did not find any matches.

Installing OpenJDKJRE64 still didnt solve it completely, but after some tries I FOUND THE SOLUTION, which is not part of the previous thread.

Usually I keep the portable applications really portable, which means not even part of the PortableApps application menu. I just run the .paf.exe directly... extract it to a new directory... and run the program from there.

if i install the app from portableapps menu, and run it from there:
the file \Data\settings\FreeplanePortableSettings.ini will contain
`PAL:LastPortableAppsBaseDir=C:\Programs\PortableApps`

which is correct. This is the value that should be there in order for the program to run properly.

But if i install the app by extracting the files and running the FreeplanePortable.exe directly, the same file will contain:
`PAL:LastPortableAppsBaseDir=C:\Programs`
or sometimes even just:
`PAL:LastPortableAppsBaseDir=C:\`
and then the error will appear and freeplane will not run.

This is the cause of the issue.
This is a small bug and should be fixed in Freeplane @dpolivaev 

so:
1. ~~Changing this value to the correct value (and installing OpenJDKJRE64) solves the issue.~~ Actually its not solving the issue because the file is being recreated every time Freeplane is running, so use the other solution.
2. Alternatively, setting the path in the file \App\AppInfo\Launcher\FreeplanePortable.ini which contains the line:
FREEPLANE_JAVA_HOME=%PAL:PortableAppsBaseDir%\PortableApps\CommonFiles\OpenJDKJRE64
replacing it an absolute path also solves the issue. Either to OpenJDKJRE64 or to jdk-17.0.3.1, both worked for me.

**Reply by dpolivaev** [reply]
What is the advantage of using a portable app? If you don't use the portable app launcher use of the paf.exe does not seem to make any sense. You can download the zip package, extract it whenever you like, run freeplane.exe and use environment variable FREEPLANE_JAVA_HOME to  help Freeplane to localize the proper Java runtime.

**Reply by sifran-github** [reply]
It is truly portable if i can run it from any directory.... not only from portableapps...
if an application provides a portable version outside of the portableapps ecosystem, then yeah I might use it.

The reason I use the portableapps versions of apps is because i know that they are configured to run and save the data within a given directory, instead of writing to global system resources like registry or files.

Maybe in the case of Freeplane, I can run it from a zip file, like u said, and it will not write to global system resources (is it true?), but most programs do write there if the user does not use a portable or portableapps version.

Sometimes I do run the programs from the portableapps menu, but sometimes I just install the .pdf.exe into a directory and just run it from there. This is what I tried to do with Freeplane.

**Comment by euu2021** [comment]
For future reference:
Easy solution, using the zip version:

- Install the version that comes with Java. It looks like this: `Freeplane-Setup-with-Java-1.11.8-pre01.exe`. The purpose here is to grab the folder "runtime", that lives inside the FreePlane installation folder. Save that "runtime" folder in a handy place, for future use. So, this step is only needed once, because, after that, you will have the "runtime" folder saved.
- Unzip the portable version normally. The version that comes in a zip, like this: `freeplane_bin-1.11.8-pre01.zip`
- Paste the "runtime" folder inside the portable version folder

**Reply by euu2021** [reply]
For future reference:
An even easier solution, using the zip version:

- google search "java portable" and you will be able to download the Java folder ready to be used;
- inside the FP folder (ie, the unzipped contents of the zip version) create a folder called "runtime";
- move the "java portable" folder contents inside this runtime folder.
